python/ray/workflow/tests/test_event_resume_after_crash.py

```diff
@@ -0,0 +1,134 @@
+import asyncio
+import subprocess
+
+from time import sleep
+import pytest
+
+import ray
+from ray import workflow, serve
+from ray.workflow.http_event_provider import HTTPListener, WorkflowEventHandleError
+from ray.tests.conftest import *  # noqa
+from ray.workflow.tests import utils
+from ray.workflow.common import WorkflowStatus
+from ray.workflow import common, workflow_context
+
+import requests
+
+
+@pytest.mark.parametrize(
+    "workflow_start_regular_shared_serve",
+    [
+        {
+            "num_cpus": 4,  # TODO (Alex): When we switch to the efficient event
+            # implementation we shouldn't need these extra cpus.
+        }
+    ],
+    indirect=True,
+)
+def test_cluster_crash_before_checkpoint(workflow_start_regular_shared_serve):
+    """If the cluster crashed before the event was checkpointed, after the cluster restarted
+    and the workflow resumed, the new event message is processed by the workflow.
+    """
+
+    class CustomHTTPListener(HTTPListener):
+        async def poll_for_event(self, event_key):
+            workflow_id = workflow_context.get_current_workflow_id()
+            if event_key is None:
+                raise WorkflowEventHandleError(
+                    workflow_id, "poll_for_event() needs event_key"
+                )
+            payload = await self.handle.get_event_payload.remote(workflow_id, event_key)
+
+            if utils.check_global_mark("after_cluster_restarted"):
+                return payload
+            else:
+                utils.set_global_mark("simulate_cluster_crash")
+                await asyncio.sleep(10000)
+
+    from ray._private import storage
+
+    storage_uri = storage._storage_uri
+
+    def send_event(msg):
+        try:
+            resp = requests.post(
+                "http://127.0.0.1:8000/event/send_event/workflow_test_cluster_crash_before_checkpoint",
+                json={"event_key": "event_key", "event_payload": msg},
+                timeout=5,
+            )
+            return resp
+        except requests.Timeout:
+            return 500
+
+    event_promise = workflow.wait_for_event(CustomHTTPListener, event_key="event_key")
+    workflow.run_async(
+        event_promise, workflow_id="workflow_test_cluster_crash_before_checkpoint"
+    )
+
+    # wait until HTTPEventProvider is ready
+    while len(serve.list_deployments().keys()) < 1:
+        sleep(0.1)
+
+    test_msg = "first_try"
+
+    while True:
+        res = send_event(test_msg)
```

Comment:
why don't you just try catch.
Also, if it timeout, why break?

I'm kind of confusing about what we are doing here.

Reply:
How about timeout then? In case of a timeout, it just break, right?